### ISSUE TITLE
Allow easier access to neighboring pixels in Dynamic module

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#201]
Currently the Dynamic module runs on each pixel, one by one:

https://github.com/publiclab/image-sequencer/blob/efd5bf7624420583097fb4ce4f83accdbb067dda/src/modules/Dynamic/Module.js#L37-L45

It has read access to the existing pixel, and uses that data to generate a new pixel color.

But to do some activities, it's helpful to be able to access neighboring pixels. This can become computationally intensive, but also useful for some operations. We could make a `neighbor(x,y)` pixel available to specify pixels `x, y` away from the current pixel.

**Comment by jywarren** [comment]
[GitMate.io](https://gitmate.io) thinks the contributor most likely able to help you is @ccpandhare.

Possibly related issues are https://github.com/publiclab/image-sequencer/issues/108 (Math Module), https://github.com/publiclab/image-sequencer/issues/146 (Monochrome "fallback" on Dynamic module not working), https://github.com/publiclab/image-sequencer/issues/112 (Scale module), https://github.com/publiclab/image-sequencer/issues/5 (image cropping module), and https://github.com/publiclab/image-sequencer/issues/75 (Pixel Manipulation and GIFs).

**Comment by tech4GT** [comment]
@jywarren should we make a function to do this in the dynamic module?
```
function neighbour(x,y){
/*Some Definition*/
}
```

**Comment by jywarren** [comment]
Yes, i think that makes sense! :-)

**Comment by ccpandhare** [comment]
I think the neighbour function isn't really solving the underlying
problem... What if I need probably the second neighbour? Maybe we could
have a more generic function which makes code more readable

So instead of
[Ndarray(i*w + j*4 + 0), Ndarray(i*w + j*4 + 1), Ndarray(i*w + j*4 + 2),
Ndarray(i*w + j*4 + 3)]

Maybe we can just do newFunction(i,j) which will return the same array
above?

Just my thoughts though!

Or maybe this could even support range based slice queries for a set of
pixels or something?

**Comment by tech4GT** [comment]
@ccpandhare actually the neighbor function is supposed to be generic only, it will yield the pixel at x,y distance from the current pixel, as @jywarren was saying, will it be fine then?

**Comment by ccpandhare** [comment]
Yeah!

**Comment by tech4GT** [comment]
@jywarren i have made this work but i have a doubt, should we put the logic to get neighboring pixels in PixelManipulation itself and pass it to changePixel so that other modules can also use it if need be?
What i mean is that changePixel can be defined like this
```
function changePixel(r,g,b,a,neighbour){
/*some code*/
neighbour(x,y) gives the neighbouring pixel
/*some other code*/
}
```
since neighbour is last parameter, its not required hence other modules which do not use neighbors would need no change
```
function changePixel(r,g,b,a){
/*function definition*/
}

**Comment by jywarren** [comment]
I think it's possible we could expose access to `neighbor()` without passing it as a parameter, if we pass it using a closure -- my brain bends a bit thinking about it, but can't we allow `changePixel()` access? Ugh, it's been too long since I did this... 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures

Otherwise, i like your passing `neighbor()` as the 5th parameter. Thanks for thinking this through!!!

**Comment by tech4GT** [comment]
@jywarren actually the thing is that the pixels object is not available in the module.js so even if we define the logic there there is no way that we get access to pixels there so we have to pass it to pixelmanipulation, only way this can work is if we bind the neighbour function to an object inside module.js and access that object directly in changepixel, would that be fine
```
Var obj={}
Obj.fun=getneighbor(){}
Changepixel uses obj.fun
```

**Comment by tech4GT** [comment]
This obj is passed to pixelmanipulation

**Comment by tech4GT** [comment]
#204 @jywarren done without extra parameter🎉

**Comment by tech4GT** [comment]
closed in #207
